### ISSUE TITLE
"git apply" fails silently under specific conditions

### ISSUE THREAD

**Issue by slackner** [wine-compholio/wine-staging#7]
Under specific conditions "git apply" fails, but doesn't return a nonzero exit-code. This bug was reported some time ago on on the #pipelight channel by package maintainers.

Although this is not a bug in wine-compholio, but instead in "git", we should check if we can somehow workaround this issue. Upstreaming a proper bugfix could be done as a second task, but we cannot assume that all distros immediately update a to a newer version, so we still need a workaround.

Code which can be used to reproduce this issue (run from an empty directory):
http://ix.io/cBq

To summarize the specific test cases (patch always contains the path {a,b}/test.txt):

* there is no git root dir at all -> *always works*
* the file exists in the git root dir, which is also the current directory -> *works*
* current directory is a subdirectory of the git root dir, file doesn't exist -> *failure (expected)*
* current directory is a subdirectory of the git root dir, file exists -> *exitcode 0, no error message, but doesn't apply*

**Comment by slackner** [comment]
A workaround was added in 603ba80d413e0d27c8d1c22ac8dd71245756fbd0. Even though I would still consider this an upstream bug (because the exitcode is wrong) I will close this bugreport, because our users do not longer need to pay attention.

For those who want to reproduce this issue (and don't want to understand the script above), try out the following steps:

* Initialize a new git directory in `dir1/dir2`
* Create some files and commit an initial version
* Create a patch with some arbitrary changes
* Delete `dir1/dir2/.git`, and initialize a git directory in `dir1`
* Change directory to `dir1/dir2` and try to apply the patch with `git apply`

The exitcode will be zero, but no patch was applied.
